lit_gpt/model.py

```diff
@@ -290,6 +291,39 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return self.proj(x)
 
 
+class GEGLU(nn.Module):
+    """
+    Source: https://github.com/pfnet-research/deep-table/blob/237c8be8a405349ce6ab78075234c60d9bfe60b7/deep_table/nn/layers/activation.py#L22
+    License: MIT, https://github.com/pfnet-research/deep-table/blob/237c8be8a405349ce6ab78075234c60d9bfe60b7/LICENSE
+    References:
+        Shazeer et al., "GLU Variants Improve Transformer," 2020.
+        https://arxiv.org/abs/2002.05202
+    """
+
+    def geglu(self, x: Tensor) -> Tensor:
+        assert x.shape[-1] % 2 == 0
+        a, b = x.chunk(2, dim=-1)
+        return a * torch.nn.functional.gelu(b)
+
+    def forward(self, x: Tensor) -> Tensor:
+        return self.geglu(x)
```

Comment:
You can half fc_1 and fc_2 in GemmaMLP to avoid introducing this class. This matches what Keras does (see my comment in https://github.com/Lightning-AI/lit-gpt/issues/940#issuecomment-1957019540)

Reply:
This is the idea

```python
class GemmaMLP(LLaMAMLP):
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        # the intermediate size for fc_{1,2} is halved when compared to LLaMAMLP, thus implementing GeGLU
        x_fc_1 = self.fc_1(x)
        x_fc_2 = self.fc_2(x)
        x = torch.nn.functional.gelu(x_fc_1, approximate="tanh") * x_fc_2
        return self.proj(x)
```

With `intermediate_size=16384 // 2,` in the config.

I'm subclassing `LLaMAMLP` because of how similar it is.

Since there's a mismatch between HF and Keras, but we get the weights from HF, the conversion script will need to half these.

This is assuming that Keras is the correct implementation. It would be useful to see the generations produced by each of them.

Reply:
Here is a notebook to run Keras variant: https://github.com/google/generative-ai-docs/blob/main/site/en/gemma/docs/get_started.ipynb